src/app/custom-table/custom-table.component.html

```diff
@@ -0,0 +1,66 @@
+      </mat-header-cell>
+      <mat-cell *matCellDef="let row" class="select-column">
+        <mat-checkbox (click)="$event.stopPropagation()" (change)="$event ? toggleCheckbox(row) : null" [checked]="selection.isSelected(row)"></mat-checkbox>
+      </mat-cell>
+    </ng-container>
+
+    <!-- Dynamic Columns -->
+    <ng-container *ngFor="let column of displayedColumns" [matColumnDef]="column">
+      <mat-header-cell *matHeaderCellDef mat-sort-header [ngClass]="column"> {{column}} </mat-header-cell>
+      <mat-cell *matCellDef="let element; trackBy: trackByFn" [ngClass]="column" (click)="startEditing(element.id)">
+        <ng-container *ngIf="editingRowIndex === element.id">
+          <ng-container *ngIf="column !== 'image' && column !== 'address'">
+            <input [(ngModel)]="editingRow[column]"  (ngModelChange)="onInputChange($event, column)" class="input" />
+          </ng-container>
+          <ng-container *ngIf="column === 'address'">
+            <div *ngFor="let key of getAddressKeys(element[column])" >
+              <label>{{key}}:</label>
+              <input [(ngModel)]="editingRow[column][key]" (ngModelChange)="onInputChange($event, column, key)" class="input" />
+            </div>
+          </ng-container>
+          <ng-container *ngIf="column === 'image'">
+            <div class="uploader-container">
+              <div appUpload (onFileDropped)="uploadFile($event)" class="file-uploader">
+                Drop a file here
+              </div>
+              <input type="file" (change)="onFileSelected($event, editingRow, column)" [disabled]="fileUploaded" />
+              <img [src]="element[column]" alt="Product Image" width="50">
+            </div>
+          </ng-container>
+        </ng-container>
+        <ng-container *ngIf="editingRowIndex !== element.id">
+          <ng-container *ngIf="column !== 'image' && column !== 'address' && !customCellTemplates[column]">{{element[column]}}</ng-container>
+          <ng-container *ngIf="column === 'address' && !customCellTemplates[column]">{{element[column] | addressFormatter}}</ng-container>
+          <ng-container *ngIf="column === 'image'">
+            <img [src]="element[column]" alt="Product Image" width="50">
+          </ng-container>
+          <ng-container *ngIf="customCellTemplates[column]">
+            <ng-template [ngTemplateOutlet]="customCellTemplates[column]" [ngTemplateOutletContext]="{ $implicit: element }"></ng-template>
+          </ng-container>
+        </ng-container>
+      </mat-cell>
+    </ng-container>
+
+    <mat-header-row *matHeaderRowDef="displayedColumnsWithSelect"></mat-header-row>
+    <mat-row *matRowDef="let row; columns: displayedColumnsWithSelect;"></mat-row>
+  </mat-table>
+
+  <mat-paginator *ngIf="tableConfig.pagination" [length]="tableConfig.totalItems" [pageSize]="tableConfig.pageSize" [pageIndex]="currentPage" (page)="onPageChange($event)"></mat-paginator>
+</div>
+
+<div class="action-bar">
```

Comment:
Just a FYI

In this case, this is a good solution, but what if you have many more different actions?
Either the logic of the disables or their display will be different for each page.

for such cases it will be more convenient to use <ng-content></ng-content>

<div class="action-bar">
  <ng-content></ng-content>
</div>

this way you can add as many different actions as you like with specific logic for each page

The button itself can also be a separate component if the main logic is duplicated many times

src/app/pages/users-table/users-table.component.html

<app-custom-table
  [data]="data"
  [tableConfig]="tableConfig"
  [customCellTemplates]="customCellTemplates"
  [deleteAction]="deleteSelected"
  [updateAction]="updateUserAction">
  <app-delete-btn></app-delete-btn>
  <app-activate-btn></app-activate-btn>
</app-custom-table>